File: php/src/AllowList.php
<?php

declare(strict_types=1);

namespace HtmlSanitization;

use ArrayAccess;
use Countable;
use IteratorAggregate;
use LogicException;
use Traversable;

/**
 * A set-like view over one of the engine's six policy lists.
 *
 * Every operation reads or writes the engine's own set — there is no PHP
 * mirror to fall out of sync.
 *
 * @implements IteratorAggregate<int, string>
 * @implements ArrayAccess<int, string>
 */
final class AllowList implements Countable, IteratorAggregate, ArrayAccess
{
    public function __construct(private HtmlSanitizer $owner, private int $which)
    {
    }

    /** Add one item, or every item of an array. Returns $this, so calls chain. */
    public function add(string|array $item): self
    {
        $ffi = $this->owner->ffi();
        foreach ((array) $item as $one) {
            $ffi->aether_hs_embed_allow($this->owner->handle(), $this->which, (string) $one);
        }
        return $this;
    }

    /** Remove one item — the "deny" direction. */
    public function remove(string $item): self
    {
        $this->owner->ffi()->aether_hs_embed_disallow($this->owner->handle(), $this->which, $item);
        return $this;
    }

    /** Empty the list — the "start from nothing" move for a strict policy. */
    public function clear(): self
    {
        $this->owner->ffi()->aether_hs_embed_clear($this->owner->handle(), $this->which);
        return $this;
    }

    public function contains(string $item): bool
    {
        return $this->owner->ffi()
            ->aether_hs_embed_is_allowed($this->owner->handle(), $this->which, $item) !== 0;
    }

    public function count(): int
    {
        return $this->owner->ffi()->aether_hs_embed_count($this->owner->handle(), $this->which);
    }

    /** The item at $index, or "" when out of range. */
    public function at(int $index): string
    {
        $ffi = $this->owner->ffi();
        return Native::takeString(
            $ffi,
            $ffi->aether_hs_embed_item_at($this->owner->handle(), $this->which, $index)
        );
    }

    /**
     * The items, in the engine's own (unspecified but stable) order.
     *
     * @return list<string>
     */
    public function toArray(): array
    {
        $out = [];
        $n = $this->count();
        for ($i = 0; $i < $n; $i++) {
            $out[] = $this->at($i);
        }
        return $out;
    }

    /**
     * The items, sorted — the deterministic version of {@see toArray()}.
     *
     * @return list<string>
     */
    public function toSortedArray(): array
    {
        $out = $this->toArray();
        sort($out, SORT_STRING);
        return $out;
    }

    public function getIterator(): Traversable
    {
        $n = $this->count();
        for ($i = 0; $i < $n; $i++) {
            yield $i => $this->at($i);
        }
    }

    public function offsetExists(mixed $offset): bool
    {
        return \is_int($offset) && $offset >= 0 && $offset < $this->count();
    }

    public function offsetGet(mixed $offset): string
    {
        return $this->at((int) $offset);
    }

    public function offsetSet(mixed $offset, mixed $value): void
    {
        // Only the append form makes sense for a set: $list[] = 'my-widget'.
        if ($offset !== null) {
            throw new LogicException('AllowList is a set; use $list[] = $item or ->add($item)');
        }
        $this->add((string) $value);
    }

    public function offsetUnset(mixed $offset): void
    {
        throw new LogicException('AllowList is a set; use ->remove($item)');
    }

    public function __toString(): string
    {
        return '{' . implode(', ', $this->toSortedArray()) . '}';
    }
}

File: java/src/main/java/org/htmlsanitizer/AllowList.java
package org.htmlsanitizer;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

/**
 * A set-like view over one of the sanitizer core's six policy lists.
 *
 * <p>Obtained from {@link HtmlSanitizer#allowedTags()} and friends. Mutations
 * go straight through to the native sanitizer — there is no local copy.
 */
public final class AllowList implements Iterable<String> {

    private final HtmlSanitizer owner;
    private final int which;

    AllowList(HtmlSanitizer owner, int which) {
        this.owner = owner;
        this.which = which;
    }

    /** Allow {@code item}. Returns this, so calls chain. */
    public AllowList add(String item) {
        try (Arena arena = Arena.ofConfined()) {
            Native api = owner.api();
            int ignored = (int) api.allow.invokeExact(
                    owner.handle(), which, arena.allocateFrom(item == null ? "" : item));
            return this;
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    public AllowList addAll(Collection<String> items) {
        for (String i : items) add(i);
        return this;
    }

    public AllowList addAll(String... items) {
        for (String i : items) add(i);
        return this;
    }

    /** The "deny" direction — drop an entry that is currently allowed. */
    public AllowList remove(String item) {
        try (Arena arena = Arena.ofConfined()) {
            Native api = owner.api();
            int ignored = (int) api.disallow.invokeExact(
                    owner.handle(), which, arena.allocateFrom(item == null ? "" : item));
            return this;
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    public boolean contains(String item) {
        try (Arena arena = Arena.ofConfined()) {
            Native api = owner.api();
            return (int) api.isAllowed.invokeExact(
                    owner.handle(), which, arena.allocateFrom(item == null ? "" : item)) != 0;
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    /** Empty the list — the "start from nothing" move for a strict policy. */
    public AllowList clear() {
        try {
            Native api = owner.api();
            int ignored = (int) api.clear.invokeExact(owner.handle(), which);
            return this;
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    public int size() {
        try {
            return (int) owner.api().count.invokeExact(owner.handle(), which);
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    /**
     * Snapshot the list. Iteration order is unspecified but each entry appears
     * exactly once.
     */
    public List<String> toList() {
        try {
            Native api = owner.api();
            int n = (int) api.count.invokeExact(owner.handle(), which);
            List<String> out = new ArrayList<>(Math.max(n, 0));
            for (int i = 0; i < n; i++) {
                out.add(api.takeString(
                        (MemorySegment) api.itemAt.invokeExact(owner.handle(), which, i)));
            }
            return out;
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    @Override
    public Iterator<String> iterator() {
        return toList().iterator();
    }

    @Override
    public String toString() {
        List<String> v = toList();
        java.util.Collections.sort(v);
        return v.toString();
    }
}
